File: src/component/Button.java
package component;

import org.newdawn.slick.Color;
import org.newdawn.slick.Font;
import org.newdawn.slick.Graphics;
import org.newdawn.slick.SlickException;
import org.newdawn.slick.gui.GUIContext;

import component.sprite.Sprite;

import core.ConstantStore;
import core.SoundStore;

/**
 * {@code Button} inherits from {@code Component} to extend features
 * that provides necessary functionality to behave like a button.
 */
public class Button extends Component implements Disableable {
	private Label label;
	private Sprite sprite;
	
	private Color buttonColor;
	private Color buttonActiveColor;
	private Color buttonDisabledColor;
	
	private boolean active;
	private boolean disabled;
	
	/**
	 * Constructs a {@code Button} with a width, a height, and a {@code Label}.
	 * @param context The GUI context
	 * @param width The width
	 * @param height The height
	 */
	public Button(GUIContext context, int width, int height) {
		super(context, width, height);
		
		buttonColor = ConstantStore.COLORS.get("INTERACTIVE_NORMAL");
		buttonActiveColor = ConstantStore.COLORS.get("INTERACTIVE_ACTIVE");
		buttonDisabledColor = ConstantStore.COLORS.get("INTERACTIVE_DISABLED");
		
		setBackgroundColor(buttonColor);
		setBevel(BevelType.OUT);
		setBevelWidth(2);
		setBorderColor(Color.black);
		setBorderWidth(2);
	}
	
	/**
	 * Constructs a {@code Button} with a width, a height, and a {@code Label}.
	 * @param context The GUI context
	 * @param width The width
	 * @param height The height
	 * @param label A label
	 */
	public Button(GUIContext context, int width, int height, Label label) {
		this(context, width, height);
		
		setLabel(label);
	}
	/**
	 * Constructs a {@code Button} with a width, a height, and a {@code Label}.
	 * @param context The GUI context
	 * @param width The width
	 * @param height The height
	 * @param label A label
	 */
	public Button(GUIContext context, int width, int height, Color color) {
		this(context, width, height);
		
		setLabel(label);
		setBackgroundColor(color);
		
	}
	
	/**
	 * Constructs a {@code Button} with a width, a height, a {@code Sprite},
	 * and a {@code Label}.
	 * @param context The GUI context
	 * @param width The width
	 * @param height The height
	 * @param sprite A sprite
	 * @param label A label
	 */
	public Button(GUIContext context, int width, int height, Sprite sprite, Label label) {
		this(context, width, height);
		
		setSprite(sprite);
		setLabel(label);
	}
	
	/**
	 * Generates the layout.
	 */
	public void layout() {
		if (label != null && label.isVisible() && sprite != null && sprite.isVisible()) {
			sprite.setPosition(this.getPosition(ReferencePoint.CENTERCENTER), ReferencePoint.BOTTOMCENTER);
			label.setPosition(this.getPosition(ReferencePoint.CENTERCENTER), ReferencePoint.TOPCENTER);
		} else {
			if (label != null) {
				label.setPosition(this.getPosition(ReferencePoint.CENTERCENTER), ReferencePoint.CENTERCENTER);
			}
			
			if (sprite != null) {
				sprite.setPosition(this.getPosition(ReferencePoint.CENTERCENTER), ReferencePoint.CENTERCENTER);
			}
		}
	}
	
	/**
	 * Sets the {@code Label}.
	 * @param label New Label to use
	 */
	public void setLabel(Label label) {
		if (this.label != null) {
			remove(this.label);
		}
		
		this.label = label;
		
		if (label != null) {
			add(label, getPosition(Positionable.ReferencePoint.CENTERCENTER), Positionable.ReferencePoint.CENTERCENTER);
		}
		
		layout();
	}
	
	/**
	 * Sets the {@code Sprite}.
	 * @param sprite New Sprite to use
	 */
	public void setSprite(Sprite sprite) {
		if (this.sprite != null) {
			remove(this.sprite);
		}
		
		this.sprite = sprite;
		
		if (sprite != null) {
			add(sprite, getPosition(Positionable.ReferencePoint.CENTERCENTER), Positionable.ReferencePoint.CENTERCENTER);
		}
		
		layout();
	}
	
	/**
	 * Sets whether or not the {@code Label} is shown.
	 * @param showLabel Whether or not the label is shown (true is visible)
	 */
	public void setShowLabel(boolean showLabel) {
		if (this.label == null) {
			return;
		}
		
		label.setVisible(showLabel);
		
		layout();
	}
	
	/**
	 * Sets whether or not the {@code Sprite} is shown.
	 * @param showSprite Whether or not the Sprite is shown (true is visible)
	 */
	public void setShowSprite(boolean showSprite) {
		if (this.sprite == null) {
			return;
		}
		
		sprite.setVisible(showSprite);
		
		layout();
	}

	@Override
	public void render(GUIContext container, Graphics g) throws SlickException {
		if (!isVisible()) {
			return;
		}
		
		super.render(container, g);
	}
	
	@Override
	public void mouseMoved(int oldx, int oldy, int newx, int newy) {
		if (!isVisible() || !isAcceptingInput()) {
			return;
		}
		
		super.mouseMoved(oldx, oldy, newx, newy);
	}
	
	@Override
	public void mousePressed(int button, int mx, int my) {
		if (!isVisible() || !isAcceptingInput()) {
			return;
		}
		
		super.mousePressed(button, mx, my);
		
		if (button == 0 && isMouseOver() && !disabled) {
			setActive(true);
			input.consumeEvent();
		}
	}
	
	@Override
	public void mouseReleased(int button, int mx, int my) {
		if (!isVisible() || !isAcceptingInput()) {
			return;
		}
		
		super.mouseReleased(button, mx, my);
				
 		if (button == 0 && isMouseOver() && !disabled && active) {
 			SoundStore.get().playSound("Click");
			notifyListeners();
			input.consumeEvent();
		}
 		
		setActive(false);
	}
	
	/**
	 * Sets the color of the button's normal background.
	 * @param color New color for the button
	 */
	public void setButtonColor(Color color) {
		buttonColor = color;
		
		// Change current color now if not disabled nor active
		if (!disabled && !active) {
			setBackgroundColor(buttonColor);
		}
	}
	
	/**
	 * Sets the color of the button's background for the active state.
	 * @param color New color for the button in the active state
	 */
	public void setButtonActiveColor(Color color) {
		buttonActiveColor = color;
	}

	/**
	 * Sets the color of the button's background for the disabled state.
	 * @param color New color for the button in the disabled state
	 */
	public void setButtonDisabledColor(Color color) {
		buttonDisabledColor = color;
	}
	
	/**
	 * Sets the font of the label.
	 * @param font The new font for the label
	 */
	public void setFont(Font font) {
		if (label == null) {
			return;
		}
		
		label.setFont(font);
	}
	
	/**
	 * Sets the color of the label.
	 * @param color The new color for the label
	 */
	public void setLabelColor(Color color) {
		if (label == null) {
			return;
		}
		
		label.setColor(color);
	}
	
	/**
	 * Sets the text of this button.
	 * @param text New text
	 */
	public void setText(String text) {
		if (label == null) {
			return;
		}
		
		label.setText(text);
	}
	
	/**
	 * Gets the text in this button.
	 * @return Text in the button
	 */
	public String getText() {
		if (label == null) {
			return "";
		}
		
		return label.getText();
	}
	
	@Override
	public void setDisabled(boolean disabled) {
		this.disabled = disabled;
		
		if (disabled) {
			this.active = false;
			
			setBackgroundColor(buttonDisabledColor);
			setBevel(BevelType.NONE);
			
			if (label != null) {
				label.setColor(ConstantStore.COLORS.get("INTERACTIVE_LABEL_DISABLED"));
			}
		} else {
			if (isActive()) {
				setBackgroundColor(buttonActiveColor);
				setBevel(BevelType.IN);
			} else {
				setBackgroundColor(buttonColor);
				setBevel(BevelType.OUT);
			}
			
			if (label != null) {
				label.setColor(ConstantStore.COLORS.get("INTERACTIVE_LABEL_NORMAL"));
			}
		}
	}
	
	@Override
	public boolean isDisabled() {
		return disabled;
	}
	
	/**
	 * Return the button's active status.
	 * @return If the button is active or not
	 */
	public boolean isActive() {
		return active;
	}
	
	/**
	 * Change whether this button is active or not.
	 * @param active New active state
	 */
	public void setActive(boolean active) {
		if (isDisabled()) {
			return;
		}
		
		this.active = active;
		
		if (active) {
			setBackgroundColor(buttonActiveColor);
			setBevel(BevelType.IN);
		} else {
			setBackgroundColor(buttonColor);
			setBevel(BevelType.OUT);
		}
	}
	
	@Override
	public String toString() {
		return "Button[active: " + active +", label: " + label + "]";
	}
}

File: web/src/component/Button.ts
import { Component, BevelType, ReferencePoint } from './Component';
import { Label } from './Label';
import { Sprite } from './Sprite';
import { Color } from '../core/Color';
import { COLORS } from '../core/ConstantStore';
import { SoundStore } from '../core/SoundStore';

export interface Disableable {
  isDisabled(): boolean;
  setDisabled(disabled: boolean): void;
}

export type ButtonClickListener = () => void;

export class Button extends Component implements Disableable {
  protected label: Label | null = null;
  protected sprite: Sprite | null = null;

  private buttonColor: Color;
  private buttonActiveColor: Color;
  private buttonDisabledColor: Color;

  private active: boolean = false;
  private disabled: boolean = false;

  private listeners: ButtonClickListener[] = [];

  constructor(width: number, height: number, labelOrColor?: Label | Color, sprite?: Sprite) {
    super(width, height);

    this.buttonColor = COLORS.get('INTERACTIVE_NORMAL') || Color.gray;
    this.buttonActiveColor = COLORS.get('INTERACTIVE_ACTIVE') || Color.darkGray;
    this.buttonDisabledColor = COLORS.get('INTERACTIVE_DISABLED') || Color.darkGray;

    if (labelOrColor instanceof Color) {
      this.buttonColor = labelOrColor;
      this.buttonActiveColor = labelOrColor;
      this.buttonDisabledColor = labelOrColor;
      this.setBackgroundColor(labelOrColor);
      this.setBorderWidth(1);
    } else {
      this.setBackgroundColor(this.buttonColor);
      this.setBevel(BevelType.OUT);
      this.setBevelWidth(2);
      this.setBorderColor(Color.black);
      this.setBorderWidth(2);

      if (labelOrColor) {
        this.setLabel(labelOrColor);
      }
    }

    if (sprite) {
      this.setSprite(sprite);
    }
  }

  layout(): void {
    if (
      this.label !== null &&
      this.label.isVisible() &&
      this.sprite !== null &&
      this.sprite.isVisible()
    ) {
      this.sprite.setPosition(
        this.getPosition(ReferencePoint.CENTERCENTER),
        ReferencePoint.BOTTOMCENTER
      );
      this.label.setPosition(
        this.getPosition(ReferencePoint.CENTERCENTER),
        ReferencePoint.TOPCENTER
      );
    } else {
      if (this.label !== null) {
        this.label.setPosition(
          this.getPosition(ReferencePoint.CENTERCENTER),
          ReferencePoint.CENTERCENTER
        );
      }

      if (this.sprite !== null) {
        this.sprite.setPosition(
          this.getPosition(ReferencePoint.CENTERCENTER),
          ReferencePoint.CENTERCENTER
        );
      }
    }
  }

  setLabel(label: Label | null): void {
    if (this.label !== null) {
      this.remove(this.label);
    }

    this.label = label;

    if (label !== null) {
      this.add(label, this.getPosition(ReferencePoint.CENTERCENTER), ReferencePoint.CENTERCENTER);
    }

    this.layout();
  }

  setSprite(sprite: Sprite | null): void {
    if (this.sprite !== null) {
      this.remove(this.sprite);
    }

    this.sprite = sprite;

    if (sprite !== null) {
      this.add(sprite, this.getPosition(ReferencePoint.CENTERCENTER), ReferencePoint.CENTERCENTER);
    }

    this.layout();
  }

  setShowLabel(showLabel: boolean): void {
    if (this.label === null) {
      return;
    }

    this.label.setVisible(showLabel);
    this.layout();
  }

  setShowSprite(showSprite: boolean): void {
    if (this.sprite === null) {
      return;
    }

    this.sprite.setVisible(showSprite);
    this.layout();
  }

  override render(ctx: CanvasRenderingContext2D): void {
    if (!this.isVisible()) {
      return;
    }

    super.render(ctx);
  }

  override mouseMoved(oldx: number, oldy: number, newx: number, newy: number): void {
    if (!this.isVisible() || !this.isAcceptingInput()) {
      return;
    }

    super.mouseMoved(oldx, oldy, newx, newy);
  }

  override mousePressed(button: number, mx: number, my: number): void {
    if (!this.isVisible() || !this.isAcceptingInput()) {
      return;
    }

    super.mousePressed(button, mx, my);

    if (button === 0 && this.isMouseOver() && !this.disabled) {
      this.setActive(true);
    }
  }

  override mouseReleased(button: number, mx: number, my: number): void {
    if (!this.isVisible() || !this.isAcceptingInput()) {
      return;
    }

    super.mouseReleased(button, mx, my);

    if (button === 0 && this.isMouseOver() && !this.disabled && this.active) {
      this.notifyListeners();
    }

    this.setActive(false);
  }

  setButtonColor(color: Color): void {
    this.buttonColor = color;

    if (!this.disabled && !this.active) {
      this.setBackgroundColor(this.buttonColor);
    }
  }

  setButtonActiveColor(color: Color): void {
    this.buttonActiveColor = color;
  }

  setButtonDisabledColor(color: Color): void {
    this.buttonDisabledColor = color;
  }

  setFont(font: any): void {
    if (this.label === null) {
      return;
    }

    this.label.setFont(font);
  }

  setLabelColor(color: Color): void {
    if (this.label === null) {
      return;
    }

    this.label.setColor(color);
  }

  setText(text: string): void {
    if (this.label === null) {
      return;
    }

    this.label.setText(text);
  }

  getText(): string {
    if (this.label === null) {
      return '';
    }

    return this.label.getText();
  }

  setDisabled(disabled: boolean): void {
    this.disabled = disabled;

    if (disabled) {
      this.active = false;

      this.setBackgroundColor(this.buttonDisabledColor);
      this.setBevel(BevelType.NONE);

      if (this.label !== null) {
        this.label.setColor(COLORS.get('INTERACTIVE_LABEL_DISABLED') || Color.gray);
      }
    } else {
      if (this.isActive()) {
        this.setBackgroundColor(this.buttonActiveColor);
        this.setBevel(BevelType.IN);
      } else {
        this.setBackgroundColor(this.buttonColor);
        this.setBevel(BevelType.OUT);
      }

      if (this.label !== null) {
        this.label.setColor(COLORS.get('INTERACTIVE_LABEL_NORMAL') || Color.white);
      }
    }
  }

  isDisabled(): boolean {
    return this.disabled;
  }

  isActive(): boolean {
    return this.active;
  }

  setActive(active: boolean): void {
    if (this.isDisabled()) {
      return;
    }

    this.active = active;

    if (active) {
      this.setBackgroundColor(this.buttonActiveColor);
      this.setBevel(BevelType.IN);
    } else {
      this.setBackgroundColor(this.buttonColor);
      this.setBevel(BevelType.OUT);
    }
  }

  addClickListener(listener: ButtonClickListener): void {
    this.listeners.push(listener);
  }

  removeClickListener(listener: ButtonClickListener): void {
    const index = this.listeners.indexOf(listener);
    if (index > -1) {
      this.listeners.splice(index, 1);
    }
  }

  protected notifyListeners(): void {
    SoundStore.playSound('Click');
    for (const listener of this.listeners) {
      listener();
    }
  }

  toString(): string {
    return `Button[active: ${this.active}, label: ${this.label}]`;
  }
}
